File: Trapping Rain Water.java
class Solution {
    public int maxWater(int arr[]) {
        // code here
        int n = arr.length;
        int pre[] = new int[n];
        int post[] = new int[n];
        pre[0] = arr[0];
        int ans = 0;
        post[n-1] = arr[n-1];
        for(int i = 1; i < n; i++){
            pre[i] = Math.max(pre[i-1],arr[i]);
            
        }
        for(int j = n-2; j >= 0; j--){
            post[j] = Math.max(post[j+1],arr[j]);
        }
        
        for(int i = 0; i < n; i++){
            ans += Math.min(pre[i],post[i]) - arr[i];
        }
        return ans;
    }
}

File: Trapping Rain Water.cpp
class Solution {
  public:
    virtual int maxWater(vector<int> &arr){
      int n=arr.size();
      vector<int> lmax(n, 0);      
      vector<int> rmax(n, 0);
      
      lmax[0]=arr[0];
      for(int i=1; i<n; i++) lmax[i]=max(lmax[i-1], arr[i]);
      
      rmax[n-1]=arr[n-1];
      for(int i=n-2; i>=0; i--) rmax[i]=max(rmax[i+1], arr[i]);
      
      
      
      int ans=0;
      for(int i=0; i<n; i++){
        int cmin=min(lmax[i], rmax[i]);
        if(cmin>arr[i]) ans+=(cmin-arr[i]);
      }
      
      return ans;
    }
};

